RFE: Use a binary tree for large filters

drakenclimber opened the pull request:
I still need to add doxygen comments in gen_pfc.c.  It also looks like I need to update the copyright headers in a few spots with 2019.

drakenclimber commented:
@tyhicks, @tych0, @jdstrand or others - would you have time to review this?  I would like to get this into the v2.5 libseccomp release if at all possible.

Also - @jdstrand and @pcmoore (and others, of course), do you have thoughts on the interface to enable/disable the binary tree?  (Again, per this issue https://github.com/seccomp/libseccomp/issues/116, it sounds like that would be nice to have.)  Perhaps we could use an `scmp_filter_attr` for disabling it?

pcmoore commented:
/me still needs to look at this

However, in regards to the question of enabling/disabling the binary tree, I'm going to stick with my earlier comments from #116 that a filter attribute is probably the best option, although I'm open to other ideas if anyone has some they want to share.  Maybe `SCMP_FLTATR_CTL_OPTIMIZE` and have the current default take a value of '1' and the binary tree approach a value of '2' (leave '0' as a special value for now)?

drakenclimber commented:
> I'm going to stick with my earlier comments from #116 that a filter attribute is probably the best option

Thanks.  I had it in my head that a filter attribute was the best way to go about this, but I totally forgot that you had originally recommended it.  The joys of getting old ;)

drakenclimber commented:
I made two major changes:
* Rebased on top of the latest master
* Added a filter attribute to enable/disable the binary tree

Also, the tests are getting rather lengthy to run.  Should we consider adding a `--fast` or `--full` option to the test suite.

Prior to these changes:
```
### python tests were disabled for this run
$ time make check
...
Regression Test Summary
 tests run: 8030
 tests skipped: 42
 tests passed: 8030
 tests failed: 0
 tests errored: 0
============================================================
PASS: regression
=============
1 test passed
=============
Making check in doc

real    5m53.913s
user    2m28.860s
sys     2m55.914s
```

And now with the binary tree tests added in:
```
### python tests were disabled for this run
$ time make check
...
Regression Test Summary
 tests run: 8362
 tests skipped: 43
 tests passed: 8362
 tests failed: 0
 tests errored: 0
============================================================
PASS: regression
=============
1 test passed
=============
Making check in doc

real    7m21.696s
user    4m9.202s
sys     3m9.842s

```

tych0 commented on src/gen_pfc.c:
```diff
@@ -241,68 +241,181 @@ static void _gen_pfc_chain(const struct arch_def *arch,
 		p_prev = NULL;
-		p_iter = p_head;
+		p_iter = *p_head;
+		while (p_iter != NULL &&
```
The only thing I might say here is that having pfc_sys_list be an array of some type instead would enable you to use qsort() or something from the standard library. You'd have to switch to an array, re-allocing along the way, etc. to use it that way, but that might be less scary than sorting things manually with all the pointer math that's thrown in above and below here.

Alternatively, you could just copy the pointers into an array, and use qsort() with a comparator that does a deref. That might be less scary code too.

drakenclimber replied:
That's an interesting idea.

I remember when I wrote this a long time back that I wanted to keep the `_sort_syscalls_by_num()` logic as close to the original `_sort_syscalls_by_priority()` logic as possible.  Seemed reasonable at the time.

Now looking at it a year later, I agree.  It's more cryptic than I would like.

tych0 replied:
Yeah, definitely no need to block this PR on the above. But also it's a matter of taste, and it's working and tested the way it is, so :man_shrugging:

drakenclimber replied:
Thanks.  Unless others object, I'm leaning that way as well.

But, I admit I may revisit this function in the future.  I have been wanting to add some unit testing to key parts of the code (some of the `db_*()` functions for example), and this may be a good place to start.  It has a fairly simple set of inputs and the desired results are also straightforward.

pcmoore replied:
FWIW, the gen_pfc code doesn't need to be *great*, it just needs to work.  It's there primarily as a debug tool, not something that would ever be used in production (it doesn't generate any *real* filter code after all).

pcmoore commented on include/seccomp.h.in:
```diff
@@ -69,6 +69,11 @@ enum scmp_filter_attr {
 	SCMP_FLTATR_CTL_SSB = 7,	/**< disable SSB mitigation */
+	SCMP_FLTATR_CTL_OPTIMIZE = 8,	/**< filter optimization level:
+					 * 0 - currently unused
+					 * 1 - default (uses syscall priorities)
```
Describing the optimization methods in the comments is admittedly hard, but I think we can do better than "syscall priorities".  I would suggest "rules weighted by priority and complexity".

pcmoore commented on include/seccomp.h.in:
```diff
@@ -69,6 +69,11 @@ enum scmp_filter_attr {
+	SCMP_FLTATR_CTL_OPTIMIZE = 8,	/**< filter optimization level:
+					 * 0 - currently unused
+					 * 1 - default (uses syscall priorities)
+					 * 2 - binary tree
```
Similar, I might say "binary tree sorted by syscall number".

pcmoore commented on src/db.c:
```diff
@@ -1375,6 +1379,25 @@ int db_col_attr_set(struct db_filter_col *col,
+			/* unsupported */
+			rc = -EOPNOTSUPP;
+			break;
+		case 1:
```
I think we can probably simplify this:

```
switch (value) {
case 1:
case 2:
        /* no comment needed here because we want you to look in seccomp.h */
        col->attr.optimize = value;
        break;
default:
        rc = -EOPNOTSUP;
        break;
}
```

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+		/* Only use a binary tree if there are a lot of syscalls
+		 * and if the filter optimization level is set to 2 (aka
+		 * binary tree)
+		 */
```
I'm thinking that since this isn't the default, if a caller explicitly asks for the binary tree optimization we should always do the optimization - thoughts?

drakenclimber replied:
I think that makes a lot of sense.  I had this logic in there from when the code was trying to intelligently decide to use a binary tree or not.  It can be removed now.

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+		 */
+		return 0;
+
+	while(max_level < syscall_cnt) {
```
Space needed, e.g. "while (..." not "while(...".  Did `make check-syntax` not catch this?

drakenclimber replied:
I ran `make check-syntax` on an earlier version of the patchset but must have forgotten to run it on this one.  Good catch.

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+		i++;
+	}
+
+	return i;
```
Perhaps I'm missing some finer point, but can't you simply divide by two here(`return syscall_cnt / 2;`)?

drakenclimber replied:
Hmmm... I'm not sure.  I remember this function was trickier to write than I expected, and I grudgingly settled on this as it seemed the best balance of readability and speed.  I'll look through it again.

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
-static struct bpf_blk *_gen_bpf_arch(struct bpf_state *state,
-				     const struct db_filter *db,
-				     const struct db_filter *db_secondary)
+static void _sort_syscalls_by_priority(struct db_sys_list *syscalls,
```
This is some good quality bikeshedding here, but I'm not typically a fan of function-names-as-sentences, how about `_sys_priority_sort(...)`?

drakenclimber replied:
Works for me.  Sometimes I fall into the Java trap of overly describing everything ;)

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+ * Sorts syscalls by their syscall number
+ *
+ */
+static void _sort_syscalls_by_num(struct db_sys_list *syscalls,
```
Similar to above, how about `_sys_num_sort(...)`?

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+ * Wrapper function for sorting syscalls
+ *
+ */
+static void _sort_syscalls(struct db_sys_list *syscalls,
```
Same, `_sys_sort(...)`?

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+ * @param db_secondary the secondary DB
+ *
+ * Count the total number of syscalls that are in the seccomp fitler
+ */
```
If this information is important, it seems like we are better off keeping that count in the db_filter struct and keeping it up to date in the db.c functions, no?

drakenclimber replied:
I like that idea.  Should be much more performant than recalculating it every time.

drakenclimber replied:
I looked into this a little bit, and it isn't as easy as I was hoping.  For x86_64, it's pretty straightforward as the syscall count goes up by one every time a new rule is added.  But it gets tricky for x86 with its multiplexed syscalls.

It's not the most elegant, but doing the counting at bpf generation time should be foolproof and performant.  I'm afraid if we try to do it when rules are added we will see several one-off bugs in architectures with weird syscall semantics.

pcmoore replied:
We're only concerned about the number of syscalls, right?  Assuming that is the case, I think it should be as easy as having `db_rule_add()` bump a counter in the db_filter struct whenever a new syscall is added to the existing db_filter struct, yes?  The arch specific manipulations shouldn't matter since they are all going to need to call into `db_rule_add()` eventually to add a rule to the filter.

Apologies if I'm missing something very obvious :)

drakenclimber replied:
Let me give it a shot again.  I played with it late one night... to very little success :(.

I'll try it again now that I have a full cup of coffee!

drakenclimber replied:
Good call.  I did a quick prototype and it seems to work nicely.  Thanks.

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1228,246 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+	state->arch = db->arch;
+
+	/* sort the syscall list */
+	syscall_cnt = _get_syscall_cnt(db, db_secondary);
```
I might move this line down so it is closer to the `_get_bintree_levels()` call.

pcmoore commented on src/gen_bpf.c:
```diff
@@ -45,6 +46,11 @@
+/* binary tree definitions */
+#define SYSCALLS_PER_NODE		(4)
+#define INVALID_HSH			(UINT64_MAX)
+#define INVALID_SYSCALL			(UINT_MAX)
```
I might preface these with something that would indicate they are specific to the binary tree optimization, especially since we use hashing for other BPF block optimizations as well.

Maybe `BTREE_HSH_INVALID`, etc.?

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1222,111 +1228,246 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
+
+		for (i = 0; i < bintree_levels; i++) {
+			bintree_syscalls[i] = INVALID_SYSCALL;
+			bintree_hashes[i]   = INVALID_HSH;
```
I realize that `_gen_bpf_arch()` is already a bit unwieldy but I wonder if we can move some of the binary tree creation into another function, much like you did with the sort?

pcmoore commented on src/gen_bpf.c:
```diff
@@ -1346,6 +1487,83 @@ static struct bpf_blk *_gen_bpf_arch(struct bpf_state *state,
+					 */
+					bintree_hashes[j] = b_bintree->hash;
+				break;
+			}
```
That entire "build the binary tree" for loop seems like a good candidate for moving into its own function, yes/no?

pcmoore replied:
Actually, looking at the code below, I'm not thinking that maybe the entire existing "create the syscall filters and add them to block list group" should be split out into separate optimization specific functions like you did with sort.

Thoughts?

drakenclimber replied:
I agree - `_gen_bpf_arch` is way too monolithic now.

I'll see what I can do to divide it into some smaller more sane functions.

drakenclimber replied:
I was able to split `_gen_bpf_arch()` up into a few smaller functions.  I think the division makes sense, but staring at that code too long may have warped my point of view ;)

pcmoore commented:
@drakenclimber do you have a revision of these patches that haven't been pushed to your PR branch?  I'm asking because I see a few comments which are marked as resolved, but the code that can be reviewed in the GH web UI doesn't reflect the resolution ... ?

pcmoore commented:
Removing this PR from the v2.5 milestone because we are already tracking this RFE for v2.5 in #116.

drakenclimber commented:
> @drakenclimber do you have a revision of these patches that haven't been pushed to your PR branch? I'm asking because I see a few comments which are marked as resolved, but the code that can be reviewed in the GH web UI doesn't reflect the resolution ... ?

Strange.  All the changes I mentioned above should be pushed to this [branch](https://github.com/drakenclimber/libseccomp/tree/rfc-bintree-v2).  I'll rebase it on top of the current master and see if that appeases - or confuses further :) - the github GUI.

drakenclimber commented:
> @drakenclimber do you have a revision of these patches that haven't been pushed to your PR branch? I'm asking because I see a few comments which are marked as resolved, but the code that can be reviewed in the GH web UI doesn't reflect the resolution ... ?

Okay, I figured out what I did.  I put all of the changes into a temporary branch because I wasn't sure how many more comments were going to come in.  My vision was that I was going to then collate them into one updated commit rather than spamming github with a million tiny changes.  But the collated commit didn't get pushed here.  I'll make sure it's cleaned up and then push it here.  Thanks.

drakenclimber commented:
> > @drakenclimber do you have a revision of these patches that haven't been pushed to your PR branch? I'm asking because I see a few comments which are marked as resolved, but the code that can be reviewed in the GH web UI doesn't reflect the resolution ... ?
> 
> Okay, I figured out what I did. I put all of the changes into a temporary branch because I wasn't sure how many more comments were going to come in. My vision was that I was going to then collate them into one updated commit rather than spamming github with a million tiny changes. But the collated commit didn't get pushed here. I'll make sure it's cleaned up and then push it here. Thanks.

Should be good to go now.  My apologies.

pcmoore commented:
No worries, it happens.  Thanks for the updated PR.

pcmoore commented:
Tsk tsk @drakenclimber, `./tools/check-syntax` is failing parts of this patch ;)

No worries, it's minor and easily fixed while merging.

pcmoore commented:
This looks good, thanks @drakenclimber.

Assuming this `make check` passes on my laptop I'll push this to master shortly.

pcmoore commented:
Everything passed, this is now in the master branch with HEAD 38f04da8474855f6ca850cc70e389bd56dbbf2fe.

drakenclimber commented:
Awesome!!!  Thanks, @pcmoore
